### PR TITLE
DAMON: adding threads to process individual user commands

### PR DESCRIPTION
**Pull request by DamonAknuh**

### CODE REVIEW

File: backend/src/web_server.py [hunk 1]
```diff
@@ -4,48 +4,57 @@
...
         binding the socket for recieving, and calling the class listen function"""
+        time_now = time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime())
+        print("-------SERVER STARTED!-------")
+        print("Timestamp: ", time_now)
+        print("Port: " + str(self.hostname))
+        print("Port: " + str(self.port))
+        print()
+        
+        # Initialize a server administration thread for dumplog command
+        self.initializeUserThread("admin")
```
Comment (DamonAknuh) [inline backend/src/web_server.py:53]: Initialized an "admin" thread for administrative tasks down the line, and for dumplog commands.

File: backend/src/web_server.py [hunk 2]
```diff
@@ -4,48 +4,57 @@
...
 
 class webServer():
 
     def __init__(self, port=SERV_PORT, hostname=SERV_HOST_NAME):
-        self.port       = port
-        self.hostname   = hostname
+        self.port           = port
+        self.hostname       = hostname
+        self.userProcesses  = {}
+        self.serverRunning  = True
```
Comment (DamonAknuh) [inline backend/src/web_server.py:39]: Added this flag to try and get server to trip out on some external input.

not really working.

File: backend/src/web_server.py [hunk 3]
```diff
@@ -62,37 +71,97 @@ def shutdown(self):
...
+        """ Every client has an asociated user thread. This thread will handle all of a specific user requests.
+            This function will initialize and start the user thread, including its queue, and user ports. """
+        
+        print("NEW USER! ", userId)
+        # Initialize the threadContext associated with the user thread. 
+        self.userProcesses[userId] = {
+            "userId" : userId,
+            "workQ"  : queue.Queue(),
+            "qLock"  : Lock(),
+        }
```
Comment (DamonAknuh) [inline backend/src/web_server.py:142]: Above is the thread context associated with each thread.  represents the information needed by each thread for the processing of commands. 

This is indexed by userId. so all users have a unique thread context, and thread.  Additional user information could be stored here down the line.

File: backend/src/web_server.py [hunk 4]
```diff
@@ -62,37 +71,97 @@ def shutdown(self):
...
+            user = strData["user"]
+
+            if user not in self.userProcesses:
+                #IF no user process thread created initialize user process thread.
+                self.initializeUserThread(user)
+
+            threadContext = self.userProcesses[user]
+
+            # Add user command packet to thread work queue
+            threadContext["workQ"].put(strData)
```
Comment (DamonAknuh) [inline backend/src/web_server.py:121]: put the user command dictionary into the queue for the user thread to process upon.

File: backend/src/web_server.py [hunk 5]
```diff
@@ -62,37 +71,97 @@ def shutdown(self):
...
+            to execute on. """
+
+        print("Starting Thread Process for Client: ", userId)
+
+        threadContext = self.userProcesses[userId]
+
+        # TODO: allow thread closing here through threadContext. 
+        while self.serverRunning:
+            # Wait for next work Q item.
+            userReq = threadContext["workQ"].get()
```
Comment (DamonAknuh) [inline backend/src/web_server.py:162]: Retrieve user command dictionary from the thread specific queue in the thread context.

File: backend/src/web_server.py [hunk 6]
```diff
@@ -62,37 +71,97 @@ def shutdown(self):
...
+            command = userReq["command"]
+            userCommands[command](userReq)
+
+        print("Ending Thread Process for Client: ", userId)
 
 
 if __name__ == '__main__':
-    server = webServer()
+    host_adr    = input("Enter hostname (localhost default): ") or "localhost"
+    port        = int(input("Enter port number (65432 default): ") or 65432)
```
Comment (DamonAknuh) [inline backend/src/web_server.py:176]: added inputs to specify server information at the beginning. Makes it easier to set port numbers and IP address than hardcoding in program.

**Review by DanielZenk — APPROVED** [review]

**Review by skosman — APPROVED** [review]

**Pull request merged**